{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Julin\n",
    "\n",
    "* Import the \"Tomato.csv\" data set from this repository\n",
    "* Make a new data frame that retains the \"trt\", \"hyp\", and \"species\" columns.\n",
    "* Now subset the new data frame do that it only has the S. chilense data.\n",
    "* Calculate the mean of the hyp column separately for the \"H\" and \"L\" treatments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shelf</th>\n",
       "      <th>flat</th>\n",
       "      <th>col</th>\n",
       "      <th>row</th>\n",
       "      <th>acs</th>\n",
       "      <th>trt</th>\n",
       "      <th>days</th>\n",
       "      <th>date</th>\n",
       "      <th>hyp</th>\n",
       "      <th>int1</th>\n",
       "      <th>...</th>\n",
       "      <th>petleng</th>\n",
       "      <th>leafleng</th>\n",
       "      <th>leafwid</th>\n",
       "      <th>leafnum</th>\n",
       "      <th>ndvi</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>alt</th>\n",
       "      <th>species</th>\n",
       "      <th>who</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1</td>\n",
       "      <td>LA2580</td>\n",
       "      <td>H</td>\n",
       "      <td>28</td>\n",
       "      <td>5/5/08</td>\n",
       "      <td>19.46</td>\n",
       "      <td>2.37</td>\n",
       "      <td>...</td>\n",
       "      <td>15.78</td>\n",
       "      <td>30.53</td>\n",
       "      <td>34.44</td>\n",
       "      <td>5.0</td>\n",
       "      <td>111</td>\n",
       "      <td>-9.5167</td>\n",
       "      <td>-78.0083</td>\n",
       "      <td>740</td>\n",
       "      <td>S. pennellii</td>\n",
       "      <td>Dan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>LA1305</td>\n",
       "      <td>H</td>\n",
       "      <td>28</td>\n",
       "      <td>5/5/08</td>\n",
       "      <td>31.28</td>\n",
       "      <td>3.34</td>\n",
       "      <td>...</td>\n",
       "      <td>12.36</td>\n",
       "      <td>22.93</td>\n",
       "      <td>13.99</td>\n",
       "      <td>4.0</td>\n",
       "      <td>120</td>\n",
       "      <td>-13.3833</td>\n",
       "      <td>-75.3583</td>\n",
       "      <td>3360</td>\n",
       "      <td>S. peruvianum</td>\n",
       "      <td>Dan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>1</td>\n",
       "      <td>LA1973</td>\n",
       "      <td>H</td>\n",
       "      <td>28</td>\n",
       "      <td>5/5/08</td>\n",
       "      <td>56.65</td>\n",
       "      <td>8.43</td>\n",
       "      <td>...</td>\n",
       "      <td>13.05</td>\n",
       "      <td>46.71</td>\n",
       "      <td>43.78</td>\n",
       "      <td>5.0</td>\n",
       "      <td>110</td>\n",
       "      <td>-16.2333</td>\n",
       "      <td>-71.7000</td>\n",
       "      <td>2585</td>\n",
       "      <td>S. peruvianum</td>\n",
       "      <td>Dan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>LA2748</td>\n",
       "      <td>H</td>\n",
       "      <td>28</td>\n",
       "      <td>5/5/08</td>\n",
       "      <td>35.18</td>\n",
       "      <td>0.56</td>\n",
       "      <td>...</td>\n",
       "      <td>8.08</td>\n",
       "      <td>26.82</td>\n",
       "      <td>33.28</td>\n",
       "      <td>5.0</td>\n",
       "      <td>105</td>\n",
       "      <td>-20.4833</td>\n",
       "      <td>-69.9833</td>\n",
       "      <td>1020</td>\n",
       "      <td>S. chilense</td>\n",
       "      <td>Dan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>LA2931</td>\n",
       "      <td>H</td>\n",
       "      <td>28</td>\n",
       "      <td>5/5/08</td>\n",
       "      <td>35.32</td>\n",
       "      <td>0.82</td>\n",
       "      <td>...</td>\n",
       "      <td>7.68</td>\n",
       "      <td>22.40</td>\n",
       "      <td>23.61</td>\n",
       "      <td>5.0</td>\n",
       "      <td>106</td>\n",
       "      <td>-20.9167</td>\n",
       "      <td>-69.0667</td>\n",
       "      <td>2460</td>\n",
       "      <td>S. chilense</td>\n",
       "      <td>Dan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  shelf  flat col  row     acs trt  days    date    hyp  int1 ...   petleng  \\\n",
       "0     Z     1   B    1  LA2580   H    28  5/5/08  19.46  2.37 ...     15.78   \n",
       "1     Z     1   C    1  LA1305   H    28  5/5/08  31.28  3.34 ...     12.36   \n",
       "2     Z     1   D    1  LA1973   H    28  5/5/08  56.65  8.43 ...     13.05   \n",
       "3     Z     1   E    1  LA2748   H    28  5/5/08  35.18  0.56 ...      8.08   \n",
       "4     Z     1   F    1  LA2931   H    28  5/5/08  35.32  0.82 ...      7.68   \n",
       "\n",
       "   leafleng  leafwid  leafnum  ndvi      lat      lon   alt        species  \\\n",
       "0     30.53    34.44      5.0   111  -9.5167 -78.0083   740   S. pennellii   \n",
       "1     22.93    13.99      4.0   120 -13.3833 -75.3583  3360  S. peruvianum   \n",
       "2     46.71    43.78      5.0   110 -16.2333 -71.7000  2585  S. peruvianum   \n",
       "3     26.82    33.28      5.0   105 -20.4833 -69.9833  1020    S. chilense   \n",
       "4     22.40    23.61      5.0   106 -20.9167 -69.0667  2460    S. chilense   \n",
       "\n",
       "   who  \n",
       "0  Dan  \n",
       "1  Dan  \n",
       "2  Dan  \n",
       "3  Dan  \n",
       "4  Dan  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tomato = pd.read_csv(\"https://raw.githubusercontent.com/UCD-pbio-rclub/python_problems/master/Tomato.csv\")\n",
    "tomato.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  trt    hyp        species\n",
      "0   H  19.46   S. pennellii\n",
      "1   H  31.28  S. peruvianum\n",
      "2   H  56.65  S. peruvianum\n",
      "3   H  35.18    S. chilense\n",
      "4   H  35.32    S. chilense\n",
      "\n",
      "  trt    hyp        species\n",
      "0   H  19.46   S. pennellii\n",
      "1   H  31.28  S. peruvianum\n",
      "2   H  56.65  S. peruvianum\n",
      "3   H  35.18    S. chilense\n",
      "4   H  35.32    S. chilense\n"
     ]
    }
   ],
   "source": [
    "subtomato = tomato[[\"trt\",\"hyp\",\"species\"]]\n",
    "print(subtomato.head())\n",
    "print()\n",
    "# This also works:\n",
    "print(tomato.loc[:,['trt', 'hyp', 'species']].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trt</th>\n",
       "      <th>hyp</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H</td>\n",
       "      <td>35.18</td>\n",
       "      <td>S. chilense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H</td>\n",
       "      <td>35.32</td>\n",
       "      <td>S. chilense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>H</td>\n",
       "      <td>28.22</td>\n",
       "      <td>S. chilense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>H</td>\n",
       "      <td>51.83</td>\n",
       "      <td>S. chilense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H</td>\n",
       "      <td>32.94</td>\n",
       "      <td>S. chilense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   trt    hyp      species\n",
       "3    H  35.18  S. chilense\n",
       "4    H  35.32  S. chilense\n",
       "10   H  28.22  S. chilense\n",
       "18   H  51.83  S. chilense\n",
       "19   H  32.94  S. chilense"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subChile = subtomato[subtomato.species==\"S. chilense\"]\n",
    "subChile.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Base or numpy:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31.52823529411765\n",
      "37.89361904761905\n"
     ]
    }
   ],
   "source": [
    "print (subChile[subChile.trt==\"H\"][\"hyp\"].mean())\n",
    "\n",
    "print (subChile[subChile.trt==\"L\"][\"hyp\"].mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "37.89361904761905"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(subChile[subChile.trt==\"H\"][\"hyp\"])\n",
    "np.mean(subChile[subChile.trt==\"L\"][\"hyp\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'H': 31.52823529411765, 'L': 37.89361904761905}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Dictionary comprehension\n",
    "{each: subChile[subChile.trt==each][\"hyp\"].mean() for each in np.unique(subChile.trt)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[31.52823529411765, 37.89361904761905]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## List comprehension\n",
    "[subChile[subChile.trt==each][\"hyp\"].mean() for each in np.unique(subChile.trt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{31.52823529411765, 37.89361904761905}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Set comprehension\n",
    "{np.mean(subChile[subChile.trt==each][\"hyp\"]) for each in np.unique(subChile.trt)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Min-Yao\n",
    "\n",
    "\n",
    "* Import the data that we used last week as a DataFrame.\n",
    "* We want to only focus on wyo_leaf_FPsc samples, so please slice out these samples.\n",
    "* In addition, assuming that wyo_leaf_FPsc_02_052 is our control sample, we want to only select the genes having expression level > 1 based on their expression level in wyo_leaf_FPsc_02_052. Please slice out a new DataFrame based on these criteria.\n",
    "* How many genes and samples we have in this new DataFrame?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wyo_leaf_FPsc_02_052</th>\n",
       "      <th>wyo_leaf_FPsc_02_064</th>\n",
       "      <th>wyo_leaf_FPsc_02_065</th>\n",
       "      <th>wyo_leaf_FPsc_04_141</th>\n",
       "      <th>wyo_leaf_FPsc_04_170</th>\n",
       "      <th>wyo_leaf_FPsc_04_174</th>\n",
       "      <th>wyo_leaf_FPsc_05_178</th>\n",
       "      <th>wyo_leaf_FPsc_05_181</th>\n",
       "      <th>wyo_leaf_FPsc_05_218</th>\n",
       "      <th>wyo_leaf_FPsc_07_272</th>\n",
       "      <th>...</th>\n",
       "      <th>wyo_leaf_R500_05_213</th>\n",
       "      <th>wyo_leaf_R500_07_266</th>\n",
       "      <th>wyo_leaf_R500_07_269</th>\n",
       "      <th>wyo_leaf_R500_07_278</th>\n",
       "      <th>wyo_leaf_R500_10_406</th>\n",
       "      <th>wyo_leaf_R500_10_424</th>\n",
       "      <th>wyo_leaf_R500_10_432</th>\n",
       "      <th>wyo_leaf_R500_12_485</th>\n",
       "      <th>wyo_leaf_R500_12_504</th>\n",
       "      <th>wyo_leaf_R500_12_513</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.767763</td>\n",
       "      <td>0.324099</td>\n",
       "      <td>0.419308</td>\n",
       "      <td>1.130997</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.872953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.603603</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.178038</td>\n",
       "      <td>-0.206271</td>\n",
       "      <td>-0.234504</td>\n",
       "      <td>-0.262736</td>\n",
       "      <td>-0.290969</td>\n",
       "      <td>-0.319201</td>\n",
       "      <td>-0.347434</td>\n",
       "      <td>-0.375667</td>\n",
       "      <td>-0.403899</td>\n",
       "      <td>-0.432132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15.512046</td>\n",
       "      <td>12.653405</td>\n",
       "      <td>12.234985</td>\n",
       "      <td>11.924020</td>\n",
       "      <td>16.025326</td>\n",
       "      <td>15.581199</td>\n",
       "      <td>13.297490</td>\n",
       "      <td>12.659967</td>\n",
       "      <td>17.091017</td>\n",
       "      <td>16.619065</td>\n",
       "      <td>...</td>\n",
       "      <td>17.125848</td>\n",
       "      <td>19.666861</td>\n",
       "      <td>15.605980</td>\n",
       "      <td>14.964833</td>\n",
       "      <td>18.584335</td>\n",
       "      <td>20.513813</td>\n",
       "      <td>15.808980</td>\n",
       "      <td>18.992666</td>\n",
       "      <td>17.519853</td>\n",
       "      <td>15.753666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.770761</td>\n",
       "      <td>0.982484</td>\n",
       "      <td>0.733789</td>\n",
       "      <td>0.487179</td>\n",
       "      <td>0.490331</td>\n",
       "      <td>0.736191</td>\n",
       "      <td>0.753390</td>\n",
       "      <td>0.474892</td>\n",
       "      <td>0.305467</td>\n",
       "      <td>0.562337</td>\n",
       "      <td>...</td>\n",
       "      <td>5.216560</td>\n",
       "      <td>7.662596</td>\n",
       "      <td>5.210235</td>\n",
       "      <td>8.325588</td>\n",
       "      <td>7.238362</td>\n",
       "      <td>6.849460</td>\n",
       "      <td>8.640743</td>\n",
       "      <td>8.025718</td>\n",
       "      <td>5.710802</td>\n",
       "      <td>8.206010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>35.481267</td>\n",
       "      <td>48.769046</td>\n",
       "      <td>44.468335</td>\n",
       "      <td>43.780244</td>\n",
       "      <td>37.906676</td>\n",
       "      <td>36.244005</td>\n",
       "      <td>46.465002</td>\n",
       "      <td>47.776153</td>\n",
       "      <td>49.213257</td>\n",
       "      <td>40.158852</td>\n",
       "      <td>...</td>\n",
       "      <td>35.080533</td>\n",
       "      <td>37.371672</td>\n",
       "      <td>46.622466</td>\n",
       "      <td>47.771217</td>\n",
       "      <td>40.133597</td>\n",
       "      <td>40.503499</td>\n",
       "      <td>46.102438</td>\n",
       "      <td>28.509853</td>\n",
       "      <td>37.889397</td>\n",
       "      <td>40.081839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.228422</td>\n",
       "      <td>2.106640</td>\n",
       "      <td>1.048270</td>\n",
       "      <td>1.454139</td>\n",
       "      <td>0.980662</td>\n",
       "      <td>2.473368</td>\n",
       "      <td>1.045839</td>\n",
       "      <td>1.103139</td>\n",
       "      <td>1.358107</td>\n",
       "      <td>3.481599</td>\n",
       "      <td>...</td>\n",
       "      <td>3.165919</td>\n",
       "      <td>3.574074</td>\n",
       "      <td>1.807563</td>\n",
       "      <td>3.577689</td>\n",
       "      <td>2.997783</td>\n",
       "      <td>2.085844</td>\n",
       "      <td>2.345345</td>\n",
       "      <td>3.174548</td>\n",
       "      <td>3.045761</td>\n",
       "      <td>3.680145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wyo_leaf_FPsc_02_052  wyo_leaf_FPsc_02_064  wyo_leaf_FPsc_02_065  \\\n",
       "0              0.767763              0.324099              0.419308   \n",
       "1             15.512046             12.653405             12.234985   \n",
       "2              0.770761              0.982484              0.733789   \n",
       "3             35.481267             48.769046             44.468335   \n",
       "4              1.228422              2.106640              1.048270   \n",
       "\n",
       "   wyo_leaf_FPsc_04_141  wyo_leaf_FPsc_04_170  wyo_leaf_FPsc_04_174  \\\n",
       "0              1.130997              0.000000              0.872953   \n",
       "1             11.924020             16.025326             15.581199   \n",
       "2              0.487179              0.490331              0.736191   \n",
       "3             43.780244             37.906676             36.244005   \n",
       "4              1.454139              0.980662              2.473368   \n",
       "\n",
       "   wyo_leaf_FPsc_05_178  wyo_leaf_FPsc_05_181  wyo_leaf_FPsc_05_218  \\\n",
       "0              0.000000              0.000000              0.603603   \n",
       "1             13.297490             12.659967             17.091017   \n",
       "2              0.753390              0.474892              0.305467   \n",
       "3             46.465002             47.776153             49.213257   \n",
       "4              1.045839              1.103139              1.358107   \n",
       "\n",
       "   wyo_leaf_FPsc_07_272          ...           wyo_leaf_R500_05_213  \\\n",
       "0              0.000000          ...                      -0.178038   \n",
       "1             16.619065          ...                      17.125848   \n",
       "2              0.562337          ...                       5.216560   \n",
       "3             40.158852          ...                      35.080533   \n",
       "4              3.481599          ...                       3.165919   \n",
       "\n",
       "   wyo_leaf_R500_07_266  wyo_leaf_R500_07_269  wyo_leaf_R500_07_278  \\\n",
       "0             -0.206271             -0.234504             -0.262736   \n",
       "1             19.666861             15.605980             14.964833   \n",
       "2              7.662596              5.210235              8.325588   \n",
       "3             37.371672             46.622466             47.771217   \n",
       "4              3.574074              1.807563              3.577689   \n",
       "\n",
       "   wyo_leaf_R500_10_406  wyo_leaf_R500_10_424  wyo_leaf_R500_10_432  \\\n",
       "0             -0.290969             -0.319201             -0.347434   \n",
       "1             18.584335             20.513813             15.808980   \n",
       "2              7.238362              6.849460              8.640743   \n",
       "3             40.133597             40.503499             46.102438   \n",
       "4              2.997783              2.085844              2.345345   \n",
       "\n",
       "   wyo_leaf_R500_12_485  wyo_leaf_R500_12_504  wyo_leaf_R500_12_513  \n",
       "0             -0.375667             -0.403899             -0.432132  \n",
       "1             18.992666             17.519853             15.753666  \n",
       "2              8.025718              5.710802              8.206010  \n",
       "3             28.509853             37.889397             40.081839  \n",
       "4              3.174548              3.045761              3.680145  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brapa = pd.read_csv(\"problemsData/Brapa_cpm.csv\")\n",
    "print(type(brapa))\n",
    "brapa.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, True),\n",
       " (1, True),\n",
       " (2, True),\n",
       " (3, True),\n",
       " (4, True),\n",
       " (5, True),\n",
       " (6, True),\n",
       " (7, True),\n",
       " (8, True),\n",
       " (9, True),\n",
       " (10, True),\n",
       " (11, True),\n",
       " (12, True),\n",
       " (13, True),\n",
       " (14, True),\n",
       " (15, True),\n",
       " (16, True),\n",
       " (17, True),\n",
       " (18, False),\n",
       " (19, False),\n",
       " (20, False),\n",
       " (21, False),\n",
       " (22, False),\n",
       " (23, False),\n",
       " (24, False),\n",
       " (25, False),\n",
       " (26, False),\n",
       " (27, False),\n",
       " (28, False),\n",
       " (29, False),\n",
       " (30, False),\n",
       " (31, False),\n",
       " (32, False),\n",
       " (33, False),\n",
       " (34, False),\n",
       " (35, False)]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Focus on wyo_leaf_FPsc samples\n",
    "import re\n",
    "list(enumerate([\"wyo_leaf_FPsc\" in sample for sample in list(brapa.columns)]))\n",
    "# Only the first 17 samples match the pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['wyo_leaf_FPsc_02_052', 'wyo_leaf_FPsc_02_064', 'wyo_leaf_FPsc_02_065',\n",
       "       'wyo_leaf_FPsc_04_141', 'wyo_leaf_FPsc_04_170', 'wyo_leaf_FPsc_04_174',\n",
       "       'wyo_leaf_FPsc_05_178', 'wyo_leaf_FPsc_05_181', 'wyo_leaf_FPsc_05_218',\n",
       "       'wyo_leaf_FPsc_07_272', 'wyo_leaf_FPsc_07_292', 'wyo_leaf_FPsc_07_296',\n",
       "       'wyo_leaf_FPsc_10_399', 'wyo_leaf_FPsc_10_428', 'wyo_leaf_FPsc_10_435',\n",
       "       'wyo_leaf_FPsc_12_508', 'wyo_leaf_FPsc_12_517'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brapa.iloc[:,0:17].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wyo_leaf_FPsc_02_052</th>\n",
       "      <th>wyo_leaf_FPsc_02_064</th>\n",
       "      <th>wyo_leaf_FPsc_02_065</th>\n",
       "      <th>wyo_leaf_FPsc_04_141</th>\n",
       "      <th>wyo_leaf_FPsc_04_170</th>\n",
       "      <th>wyo_leaf_FPsc_04_174</th>\n",
       "      <th>wyo_leaf_FPsc_05_178</th>\n",
       "      <th>wyo_leaf_FPsc_05_181</th>\n",
       "      <th>wyo_leaf_FPsc_05_218</th>\n",
       "      <th>wyo_leaf_FPsc_07_272</th>\n",
       "      <th>wyo_leaf_FPsc_07_292</th>\n",
       "      <th>wyo_leaf_FPsc_07_296</th>\n",
       "      <th>wyo_leaf_FPsc_10_399</th>\n",
       "      <th>wyo_leaf_FPsc_10_428</th>\n",
       "      <th>wyo_leaf_FPsc_10_435</th>\n",
       "      <th>wyo_leaf_FPsc_12_508</th>\n",
       "      <th>wyo_leaf_FPsc_12_517</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.767763</td>\n",
       "      <td>0.324099</td>\n",
       "      <td>0.419308</td>\n",
       "      <td>1.130997</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.872953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.603603</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.148906</td>\n",
       "      <td>0.302771</td>\n",
       "      <td>0.157520</td>\n",
       "      <td>0.276972</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.504064</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15.512046</td>\n",
       "      <td>12.653405</td>\n",
       "      <td>12.234985</td>\n",
       "      <td>11.924020</td>\n",
       "      <td>16.025326</td>\n",
       "      <td>15.581199</td>\n",
       "      <td>13.297490</td>\n",
       "      <td>12.659967</td>\n",
       "      <td>17.091017</td>\n",
       "      <td>16.619065</td>\n",
       "      <td>18.259009</td>\n",
       "      <td>16.139659</td>\n",
       "      <td>13.857280</td>\n",
       "      <td>13.507773</td>\n",
       "      <td>13.775680</td>\n",
       "      <td>15.408758</td>\n",
       "      <td>15.323277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.770761</td>\n",
       "      <td>0.982484</td>\n",
       "      <td>0.733789</td>\n",
       "      <td>0.487179</td>\n",
       "      <td>0.490331</td>\n",
       "      <td>0.736191</td>\n",
       "      <td>0.753390</td>\n",
       "      <td>0.474892</td>\n",
       "      <td>0.305467</td>\n",
       "      <td>0.562337</td>\n",
       "      <td>0.598514</td>\n",
       "      <td>0.306845</td>\n",
       "      <td>0.477273</td>\n",
       "      <td>0.422110</td>\n",
       "      <td>0.737009</td>\n",
       "      <td>0.337412</td>\n",
       "      <td>0.452063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>35.481267</td>\n",
       "      <td>48.769046</td>\n",
       "      <td>44.468335</td>\n",
       "      <td>43.780244</td>\n",
       "      <td>37.906676</td>\n",
       "      <td>36.244005</td>\n",
       "      <td>46.465002</td>\n",
       "      <td>47.776153</td>\n",
       "      <td>49.213257</td>\n",
       "      <td>40.158852</td>\n",
       "      <td>42.432119</td>\n",
       "      <td>44.930602</td>\n",
       "      <td>35.286235</td>\n",
       "      <td>40.941962</td>\n",
       "      <td>35.568740</td>\n",
       "      <td>31.664813</td>\n",
       "      <td>47.754837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.228422</td>\n",
       "      <td>2.106640</td>\n",
       "      <td>1.048270</td>\n",
       "      <td>1.454139</td>\n",
       "      <td>0.980662</td>\n",
       "      <td>2.473368</td>\n",
       "      <td>1.045839</td>\n",
       "      <td>1.103139</td>\n",
       "      <td>1.358107</td>\n",
       "      <td>3.481599</td>\n",
       "      <td>1.489060</td>\n",
       "      <td>1.816626</td>\n",
       "      <td>2.520322</td>\n",
       "      <td>1.107887</td>\n",
       "      <td>1.160623</td>\n",
       "      <td>1.848236</td>\n",
       "      <td>2.702280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   wyo_leaf_FPsc_02_052  wyo_leaf_FPsc_02_064  wyo_leaf_FPsc_02_065  \\\n",
       "0              0.767763              0.324099              0.419308   \n",
       "1             15.512046             12.653405             12.234985   \n",
       "2              0.770761              0.982484              0.733789   \n",
       "3             35.481267             48.769046             44.468335   \n",
       "4              1.228422              2.106640              1.048270   \n",
       "\n",
       "   wyo_leaf_FPsc_04_141  wyo_leaf_FPsc_04_170  wyo_leaf_FPsc_04_174  \\\n",
       "0              1.130997              0.000000              0.872953   \n",
       "1             11.924020             16.025326             15.581199   \n",
       "2              0.487179              0.490331              0.736191   \n",
       "3             43.780244             37.906676             36.244005   \n",
       "4              1.454139              0.980662              2.473368   \n",
       "\n",
       "   wyo_leaf_FPsc_05_178  wyo_leaf_FPsc_05_181  wyo_leaf_FPsc_05_218  \\\n",
       "0              0.000000              0.000000              0.603603   \n",
       "1             13.297490             12.659967             17.091017   \n",
       "2              0.753390              0.474892              0.305467   \n",
       "3             46.465002             47.776153             49.213257   \n",
       "4              1.045839              1.103139              1.358107   \n",
       "\n",
       "   wyo_leaf_FPsc_07_272  wyo_leaf_FPsc_07_292  wyo_leaf_FPsc_07_296  \\\n",
       "0              0.000000              0.148906              0.302771   \n",
       "1             16.619065             18.259009             16.139659   \n",
       "2              0.562337              0.598514              0.306845   \n",
       "3             40.158852             42.432119             44.930602   \n",
       "4              3.481599              1.489060              1.816626   \n",
       "\n",
       "   wyo_leaf_FPsc_10_399  wyo_leaf_FPsc_10_428  wyo_leaf_FPsc_10_435  \\\n",
       "0              0.157520              0.276972              0.000000   \n",
       "1             13.857280             13.507773             13.775680   \n",
       "2              0.477273              0.422110              0.737009   \n",
       "3             35.286235             40.941962             35.568740   \n",
       "4              2.520322              1.107887              1.160623   \n",
       "\n",
       "   wyo_leaf_FPsc_12_508  wyo_leaf_FPsc_12_517  \n",
       "0              0.504064              0.000000  \n",
       "1             15.408758             15.323277  \n",
       "2              0.337412              0.452063  \n",
       "3             31.664813             47.754837  \n",
       "4              1.848236              2.702280  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subBrapa = brapa.iloc[:,0:17]\n",
    "subBrapa.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "subBrapa_expr = subBrapa[subBrapa.loc[:,\"wyo_leaf_FPsc_02_052\"] >= 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65, 17)"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subBrapa_expr.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rie\n",
    "\n",
    "* Import Brapa_cpm.csv dataset that we used last week.\n",
    "* Extract wyo_leaf_FPsc_04_141,wyo_leaf_FPsc_04_170, wyo_leaf_FPsc_04_174 samples, assuming those samples are biological replicates.\n",
    "* Make a new data frame (I call it dataA) with these 3 samples. Get averages for each gene. Append the averages to dataA.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wyo_leaf_FPsc_02_052</th>\n",
       "      <th>wyo_leaf_FPsc_02_064</th>\n",
       "      <th>wyo_leaf_FPsc_02_065</th>\n",
       "      <th>wyo_leaf_FPsc_04_141</th>\n",
       "      <th>wyo_leaf_FPsc_04_170</th>\n",
       "      <th>wyo_leaf_FPsc_04_174</th>\n",
       "      <th>wyo_leaf_FPsc_05_178</th>\n",
       "      <th>wyo_leaf_FPsc_05_181</th>\n",
       "      <th>wyo_leaf_FPsc_05_218</th>\n",
       "      <th>wyo_leaf_FPsc_07_272</th>\n",
       "      <th>...</th>\n",
       "      <th>wyo_leaf_R500_05_213</th>\n",
       "      <th>wyo_leaf_R500_07_266</th>\n",
       "      <th>wyo_leaf_R500_07_269</th>\n",
       "      <th>wyo_leaf_R500_07_278</th>\n",
       "      <th>wyo_leaf_R500_10_406</th>\n",
       "      <th>wyo_leaf_R500_10_424</th>\n",
       "      <th>wyo_leaf_R500_10_432</th>\n",
       "      <th>wyo_leaf_R500_12_485</th>\n",
       "      <th>wyo_leaf_R500_12_504</th>\n",
       "      <th>wyo_leaf_R500_12_513</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.767763</td>\n",
       "      <td>0.324099</td>\n",
       "      <td>0.419308</td>\n",
       "      <td>1.130997</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.872953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.603603</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.178038</td>\n",
       "      <td>-0.206271</td>\n",
       "      <td>-0.234504</td>\n",
       "      <td>-0.262736</td>\n",
       "      <td>-0.290969</td>\n",
       "      <td>-0.319201</td>\n",
       "      <td>-0.347434</td>\n",
       "      <td>-0.375667</td>\n",
       "      <td>-0.403899</td>\n",
       "      <td>-0.432132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15.512046</td>\n",
       "      <td>12.653405</td>\n",
       "      <td>12.234985</td>\n",
       "      <td>11.924020</td>\n",
       "      <td>16.025326</td>\n",
       "      <td>15.581199</td>\n",
       "      <td>13.297490</td>\n",
       "      <td>12.659967</td>\n",
       "      <td>17.091017</td>\n",
       "      <td>16.619065</td>\n",
       "      <td>...</td>\n",
       "      <td>17.125848</td>\n",
       "      <td>19.666861</td>\n",
       "      <td>15.605980</td>\n",
       "      <td>14.964833</td>\n",
       "      <td>18.584335</td>\n",
       "      <td>20.513813</td>\n",
       "      <td>15.808980</td>\n",
       "      <td>18.992666</td>\n",
       "      <td>17.519853</td>\n",
       "      <td>15.753666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.770761</td>\n",
       "      <td>0.982484</td>\n",
       "      <td>0.733789</td>\n",
       "      <td>0.487179</td>\n",
       "      <td>0.490331</td>\n",
       "      <td>0.736191</td>\n",
       "      <td>0.753390</td>\n",
       "      <td>0.474892</td>\n",
       "      <td>0.305467</td>\n",
       "      <td>0.562337</td>\n",
       "      <td>...</td>\n",
       "      <td>5.216560</td>\n",
       "      <td>7.662596</td>\n",
       "      <td>5.210235</td>\n",
       "      <td>8.325588</td>\n",
       "      <td>7.238362</td>\n",
       "      <td>6.849460</td>\n",
       "      <td>8.640743</td>\n",
       "      <td>8.025718</td>\n",
       "      <td>5.710802</td>\n",
       "      <td>8.206010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>35.481267</td>\n",
       "      <td>48.769046</td>\n",
       "      <td>44.468335</td>\n",
       "      <td>43.780244</td>\n",
       "      <td>37.906676</td>\n",
       "      <td>36.244005</td>\n",
       "      <td>46.465002</td>\n",
       "      <td>47.776153</td>\n",
       "      <td>49.213257</td>\n",
       "      <td>40.158852</td>\n",
       "      <td>...</td>\n",
       "      <td>35.080533</td>\n",
       "      <td>37.371672</td>\n",
       "      <td>46.622466</td>\n",
       "      <td>47.771217</td>\n",
       "      <td>40.133597</td>\n",
       "      <td>40.503499</td>\n",
       "      <td>46.102438</td>\n",
       "      <td>28.509853</td>\n",
       "      <td>37.889397</td>\n",
       "      <td>40.081839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.228422</td>\n",
       "      <td>2.106640</td>\n",
       "      <td>1.048270</td>\n",
       "      <td>1.454139</td>\n",
       "      <td>0.980662</td>\n",
       "      <td>2.473368</td>\n",
       "      <td>1.045839</td>\n",
       "      <td>1.103139</td>\n",
       "      <td>1.358107</td>\n",
       "      <td>3.481599</td>\n",
       "      <td>...</td>\n",
       "      <td>3.165919</td>\n",
       "      <td>3.574074</td>\n",
       "      <td>1.807563</td>\n",
       "      <td>3.577689</td>\n",
       "      <td>2.997783</td>\n",
       "      <td>2.085844</td>\n",
       "      <td>2.345345</td>\n",
       "      <td>3.174548</td>\n",
       "      <td>3.045761</td>\n",
       "      <td>3.680145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wyo_leaf_FPsc_02_052  wyo_leaf_FPsc_02_064  wyo_leaf_FPsc_02_065  \\\n",
       "0              0.767763              0.324099              0.419308   \n",
       "1             15.512046             12.653405             12.234985   \n",
       "2              0.770761              0.982484              0.733789   \n",
       "3             35.481267             48.769046             44.468335   \n",
       "4              1.228422              2.106640              1.048270   \n",
       "\n",
       "   wyo_leaf_FPsc_04_141  wyo_leaf_FPsc_04_170  wyo_leaf_FPsc_04_174  \\\n",
       "0              1.130997              0.000000              0.872953   \n",
       "1             11.924020             16.025326             15.581199   \n",
       "2              0.487179              0.490331              0.736191   \n",
       "3             43.780244             37.906676             36.244005   \n",
       "4              1.454139              0.980662              2.473368   \n",
       "\n",
       "   wyo_leaf_FPsc_05_178  wyo_leaf_FPsc_05_181  wyo_leaf_FPsc_05_218  \\\n",
       "0              0.000000              0.000000              0.603603   \n",
       "1             13.297490             12.659967             17.091017   \n",
       "2              0.753390              0.474892              0.305467   \n",
       "3             46.465002             47.776153             49.213257   \n",
       "4              1.045839              1.103139              1.358107   \n",
       "\n",
       "   wyo_leaf_FPsc_07_272          ...           wyo_leaf_R500_05_213  \\\n",
       "0              0.000000          ...                      -0.178038   \n",
       "1             16.619065          ...                      17.125848   \n",
       "2              0.562337          ...                       5.216560   \n",
       "3             40.158852          ...                      35.080533   \n",
       "4              3.481599          ...                       3.165919   \n",
       "\n",
       "   wyo_leaf_R500_07_266  wyo_leaf_R500_07_269  wyo_leaf_R500_07_278  \\\n",
       "0             -0.206271             -0.234504             -0.262736   \n",
       "1             19.666861             15.605980             14.964833   \n",
       "2              7.662596              5.210235              8.325588   \n",
       "3             37.371672             46.622466             47.771217   \n",
       "4              3.574074              1.807563              3.577689   \n",
       "\n",
       "   wyo_leaf_R500_10_406  wyo_leaf_R500_10_424  wyo_leaf_R500_10_432  \\\n",
       "0             -0.290969             -0.319201             -0.347434   \n",
       "1             18.584335             20.513813             15.808980   \n",
       "2              7.238362              6.849460              8.640743   \n",
       "3             40.133597             40.503499             46.102438   \n",
       "4              2.997783              2.085844              2.345345   \n",
       "\n",
       "   wyo_leaf_R500_12_485  wyo_leaf_R500_12_504  wyo_leaf_R500_12_513  \n",
       "0             -0.375667             -0.403899             -0.432132  \n",
       "1             18.992666             17.519853             15.753666  \n",
       "2              8.025718              5.710802              8.206010  \n",
       "3             28.509853             37.889397             40.081839  \n",
       "4              3.174548              3.045761              3.680145  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R\n",
    "\n",
    "https://github.com/cuttlefishh/python-for-data-analysis/blob/master/assignments/assignment6.md"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2000, 75)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_table(\"ftp://ftp.microbio.me/emp/release1/mapping_files/emp_qiime_mapping_subset_2k.tsv\",index_col=0)\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Description</th>\n",
       "      <th>host_subject_id</th>\n",
       "      <th>study_id</th>\n",
       "      <th>title</th>\n",
       "      <th>principal_investigator</th>\n",
       "      <th>doi</th>\n",
       "      <th>ebi_accession</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>adiv_shannon</th>\n",
       "      <th>adiv_faith_pd</th>\n",
       "      <th>temperature_deg_c</th>\n",
       "      <th>ph</th>\n",
       "      <th>salinity_psu</th>\n",
       "      <th>oxygen_mg_per_l</th>\n",
       "      <th>phosphate_umol_per_l</th>\n",
       "      <th>ammonium_umol_per_l</th>\n",
       "      <th>nitrate_umol_per_l</th>\n",
       "      <th>sulfate_umol_per_l</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>550.L1S116.s.1.sequence</th>\n",
       "      <td>ATGCCTGAGCAG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>sample_20 stool</td>\n",
       "      <td>F4</td>\n",
       "      <td>550</td>\n",
       "      <td>Moving pictures of the human microbiome</td>\n",
       "      <td>Rob Knight</td>\n",
       "      <td>10.1186/gb-2011-12-5-r50</td>\n",
       "      <td>ERP021896</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>3.867414</td>\n",
       "      <td>12.457989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S119.s.1.sequence</th>\n",
       "      <td>CAGCACTAAGCG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>sample_23 stool</td>\n",
       "      <td>F4</td>\n",
       "      <td>550</td>\n",
       "      <td>Moving pictures of the human microbiome</td>\n",
       "      <td>Rob Knight</td>\n",
       "      <td>10.1186/gb-2011-12-5-r50</td>\n",
       "      <td>ERP021896</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>3.265164</td>\n",
       "      <td>10.719448</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S164.s.1.sequence</th>\n",
       "      <td>ATGTACGGCGAC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>sample_73 stool</td>\n",
       "      <td>M3</td>\n",
       "      <td>550</td>\n",
       "      <td>Moving pictures of the human microbiome</td>\n",
       "      <td>Rob Knight</td>\n",
       "      <td>10.1186/gb-2011-12-5-r50</td>\n",
       "      <td>ERP021896</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>3.661124</td>\n",
       "      <td>14.214158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        BarcodeSequence LinkerPrimerSequence      Description  \\\n",
       "#SampleID                                                                       \n",
       "550.L1S116.s.1.sequence    ATGCCTGAGCAG  GTGCCAGCMGCCGCGGTAA  sample_20 stool   \n",
       "550.L1S119.s.1.sequence    CAGCACTAAGCG  GTGCCAGCMGCCGCGGTAA  sample_23 stool   \n",
       "550.L1S164.s.1.sequence    ATGTACGGCGAC  GTGCCAGCMGCCGCGGTAA  sample_73 stool   \n",
       "\n",
       "                        host_subject_id  study_id  \\\n",
       "#SampleID                                           \n",
       "550.L1S116.s.1.sequence              F4       550   \n",
       "550.L1S119.s.1.sequence              F4       550   \n",
       "550.L1S164.s.1.sequence              M3       550   \n",
       "\n",
       "                                                           title  \\\n",
       "#SampleID                                                          \n",
       "550.L1S116.s.1.sequence  Moving pictures of the human microbiome   \n",
       "550.L1S119.s.1.sequence  Moving pictures of the human microbiome   \n",
       "550.L1S164.s.1.sequence  Moving pictures of the human microbiome   \n",
       "\n",
       "                        principal_investigator                       doi  \\\n",
       "#SampleID                                                                  \n",
       "550.L1S116.s.1.sequence             Rob Knight  10.1186/gb-2011-12-5-r50   \n",
       "550.L1S119.s.1.sequence             Rob Knight  10.1186/gb-2011-12-5-r50   \n",
       "550.L1S164.s.1.sequence             Rob Knight  10.1186/gb-2011-12-5-r50   \n",
       "\n",
       "                        ebi_accession target_gene         ...          \\\n",
       "#SampleID                                                 ...           \n",
       "550.L1S116.s.1.sequence     ERP021896    16S rRNA         ...           \n",
       "550.L1S119.s.1.sequence     ERP021896    16S rRNA         ...           \n",
       "550.L1S164.s.1.sequence     ERP021896    16S rRNA         ...           \n",
       "\n",
       "                        adiv_shannon adiv_faith_pd temperature_deg_c  ph  \\\n",
       "#SampleID                                                                  \n",
       "550.L1S116.s.1.sequence     3.867414     12.457989               NaN NaN   \n",
       "550.L1S119.s.1.sequence     3.265164     10.719448               NaN NaN   \n",
       "550.L1S164.s.1.sequence     3.661124     14.214158               NaN NaN   \n",
       "\n",
       "                        salinity_psu oxygen_mg_per_l  phosphate_umol_per_l  \\\n",
       "#SampleID                                                                    \n",
       "550.L1S116.s.1.sequence          NaN             NaN                   NaN   \n",
       "550.L1S119.s.1.sequence          NaN             NaN                   NaN   \n",
       "550.L1S164.s.1.sequence          NaN             NaN                   NaN   \n",
       "\n",
       "                         ammonium_umol_per_l  nitrate_umol_per_l  \\\n",
       "#SampleID                                                          \n",
       "550.L1S116.s.1.sequence                  NaN                 NaN   \n",
       "550.L1S119.s.1.sequence                  NaN                 NaN   \n",
       "550.L1S164.s.1.sequence                  NaN                 NaN   \n",
       "\n",
       "                         sulfate_umol_per_l  \n",
       "#SampleID                                    \n",
       "550.L1S116.s.1.sequence                 NaN  \n",
       "550.L1S119.s.1.sequence                 NaN  \n",
       "550.L1S164.s.1.sequence                 NaN  \n",
       "\n",
       "[3 rows x 75 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Reading and summarizing__\n",
    "* Import the tab-separated values file emp_qiime_mapping_subset_2k.tsv as a DataFrame called df with default data types, with the first row as column labels (columns) and the first column as row labels (indexes).\n",
    "* The indexes should be the sample IDs. How many samples are in this DataFrame? How many metadata columns?\n",
    "* What are the minimum and maximum pH values in the dataset?\n",
    "* What are the average and standard deviation temperature values in the dataset?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    411.000000\n",
       "mean      18.491088\n",
       "std       16.672381\n",
       "min      -15.000000\n",
       "25%        9.714750\n",
       "50%       15.240000\n",
       "75%       20.207798\n",
       "max       91.000000\n",
       "Name: temperature_deg_c, dtype: float64"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.temperature_deg_c.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Indexing, slicing, and writing__\n",
    "* Make a new Series called temp with the temperature column as its own Series object. Remove NaN values (np.nan) from this Series. How many values are left?\n",
    "* Make a new DataFrame called df_seqs from columns sequences_split_libraries through observations_deblur_150bp (column positions 17-23) of the existing DataFrame. What is the mean value of column observations_deblur_90bp?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.series.Series'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "#SampleID\n",
       "678.OA.mesocosm.362    11.0\n",
       "678.OA.mesocosm.376    10.8\n",
       "678.OA.mesocosm.410    12.1\n",
       "Name: temperature_deg_c, dtype: float64"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temperature = df.temperature_deg_c[~np.isnan(df.temperature_deg_c)]\n",
    "print(type(temperature))\n",
    "temperature.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequences_split_libraries</th>\n",
       "      <th>observations_closed_ref_greengenes</th>\n",
       "      <th>observations_closed_ref_silva</th>\n",
       "      <th>observations_open_ref_greengenes</th>\n",
       "      <th>observations_deblur_90bp</th>\n",
       "      <th>observations_deblur_100bp</th>\n",
       "      <th>observations_deblur_150bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>550.L1S116.s.1.sequence</th>\n",
       "      <td>33383</td>\n",
       "      <td>32153</td>\n",
       "      <td>32453</td>\n",
       "      <td>33337</td>\n",
       "      <td>22567</td>\n",
       "      <td>22160</td>\n",
       "      <td>1043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S119.s.1.sequence</th>\n",
       "      <td>40944</td>\n",
       "      <td>39472</td>\n",
       "      <td>39929</td>\n",
       "      <td>40870</td>\n",
       "      <td>27871</td>\n",
       "      <td>27191</td>\n",
       "      <td>1272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S164.s.1.sequence</th>\n",
       "      <td>35636</td>\n",
       "      <td>34550</td>\n",
       "      <td>34666</td>\n",
       "      <td>35599</td>\n",
       "      <td>24134</td>\n",
       "      <td>23686</td>\n",
       "      <td>1161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S194.s.1.sequence</th>\n",
       "      <td>46992</td>\n",
       "      <td>43925</td>\n",
       "      <td>43852</td>\n",
       "      <td>46875</td>\n",
       "      <td>30041</td>\n",
       "      <td>29264</td>\n",
       "      <td>1974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S20.s.1.sequence</th>\n",
       "      <td>30131</td>\n",
       "      <td>29179</td>\n",
       "      <td>29553</td>\n",
       "      <td>30094</td>\n",
       "      <td>21132</td>\n",
       "      <td>20643</td>\n",
       "      <td>603</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sequences_split_libraries  \\\n",
       "#SampleID                                            \n",
       "550.L1S116.s.1.sequence                      33383   \n",
       "550.L1S119.s.1.sequence                      40944   \n",
       "550.L1S164.s.1.sequence                      35636   \n",
       "550.L1S194.s.1.sequence                      46992   \n",
       "550.L1S20.s.1.sequence                       30131   \n",
       "\n",
       "                         observations_closed_ref_greengenes  \\\n",
       "#SampleID                                                     \n",
       "550.L1S116.s.1.sequence                               32153   \n",
       "550.L1S119.s.1.sequence                               39472   \n",
       "550.L1S164.s.1.sequence                               34550   \n",
       "550.L1S194.s.1.sequence                               43925   \n",
       "550.L1S20.s.1.sequence                                29179   \n",
       "\n",
       "                         observations_closed_ref_silva  \\\n",
       "#SampleID                                                \n",
       "550.L1S116.s.1.sequence                          32453   \n",
       "550.L1S119.s.1.sequence                          39929   \n",
       "550.L1S164.s.1.sequence                          34666   \n",
       "550.L1S194.s.1.sequence                          43852   \n",
       "550.L1S20.s.1.sequence                           29553   \n",
       "\n",
       "                         observations_open_ref_greengenes  \\\n",
       "#SampleID                                                   \n",
       "550.L1S116.s.1.sequence                             33337   \n",
       "550.L1S119.s.1.sequence                             40870   \n",
       "550.L1S164.s.1.sequence                             35599   \n",
       "550.L1S194.s.1.sequence                             46875   \n",
       "550.L1S20.s.1.sequence                              30094   \n",
       "\n",
       "                         observations_deblur_90bp  observations_deblur_100bp  \\\n",
       "#SampleID                                                                      \n",
       "550.L1S116.s.1.sequence                     22567                      22160   \n",
       "550.L1S119.s.1.sequence                     27871                      27191   \n",
       "550.L1S164.s.1.sequence                     24134                      23686   \n",
       "550.L1S194.s.1.sequence                     30041                      29264   \n",
       "550.L1S20.s.1.sequence                      21132                      20643   \n",
       "\n",
       "                         observations_deblur_150bp  \n",
       "#SampleID                                           \n",
       "550.L1S116.s.1.sequence                       1043  \n",
       "550.L1S119.s.1.sequence                       1272  \n",
       "550.L1S164.s.1.sequence                       1161  \n",
       "550.L1S194.s.1.sequence                       1974  \n",
       "550.L1S20.s.1.sequence                         603  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seqs=df.loc[:,\"sequences_split_libraries\":\"observations_deblur_150bp\"]\n",
    "df_seqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "cell_style": "split"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequences_split_libraries</th>\n",
       "      <th>observations_closed_ref_greengenes</th>\n",
       "      <th>observations_closed_ref_silva</th>\n",
       "      <th>observations_open_ref_greengenes</th>\n",
       "      <th>observations_deblur_90bp</th>\n",
       "      <th>observations_deblur_100bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>550.L1S116.s.1.sequence</th>\n",
       "      <td>33383</td>\n",
       "      <td>32153</td>\n",
       "      <td>32453</td>\n",
       "      <td>33337</td>\n",
       "      <td>22567</td>\n",
       "      <td>22160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S119.s.1.sequence</th>\n",
       "      <td>40944</td>\n",
       "      <td>39472</td>\n",
       "      <td>39929</td>\n",
       "      <td>40870</td>\n",
       "      <td>27871</td>\n",
       "      <td>27191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S164.s.1.sequence</th>\n",
       "      <td>35636</td>\n",
       "      <td>34550</td>\n",
       "      <td>34666</td>\n",
       "      <td>35599</td>\n",
       "      <td>24134</td>\n",
       "      <td>23686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S194.s.1.sequence</th>\n",
       "      <td>46992</td>\n",
       "      <td>43925</td>\n",
       "      <td>43852</td>\n",
       "      <td>46875</td>\n",
       "      <td>30041</td>\n",
       "      <td>29264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550.L1S20.s.1.sequence</th>\n",
       "      <td>30131</td>\n",
       "      <td>29179</td>\n",
       "      <td>29553</td>\n",
       "      <td>30094</td>\n",
       "      <td>21132</td>\n",
       "      <td>20643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sequences_split_libraries  \\\n",
       "#SampleID                                            \n",
       "550.L1S116.s.1.sequence                      33383   \n",
       "550.L1S119.s.1.sequence                      40944   \n",
       "550.L1S164.s.1.sequence                      35636   \n",
       "550.L1S194.s.1.sequence                      46992   \n",
       "550.L1S20.s.1.sequence                       30131   \n",
       "\n",
       "                         observations_closed_ref_greengenes  \\\n",
       "#SampleID                                                     \n",
       "550.L1S116.s.1.sequence                               32153   \n",
       "550.L1S119.s.1.sequence                               39472   \n",
       "550.L1S164.s.1.sequence                               34550   \n",
       "550.L1S194.s.1.sequence                               43925   \n",
       "550.L1S20.s.1.sequence                                29179   \n",
       "\n",
       "                         observations_closed_ref_silva  \\\n",
       "#SampleID                                                \n",
       "550.L1S116.s.1.sequence                          32453   \n",
       "550.L1S119.s.1.sequence                          39929   \n",
       "550.L1S164.s.1.sequence                          34666   \n",
       "550.L1S194.s.1.sequence                          43852   \n",
       "550.L1S20.s.1.sequence                           29553   \n",
       "\n",
       "                         observations_open_ref_greengenes  \\\n",
       "#SampleID                                                   \n",
       "550.L1S116.s.1.sequence                             33337   \n",
       "550.L1S119.s.1.sequence                             40870   \n",
       "550.L1S164.s.1.sequence                             35599   \n",
       "550.L1S194.s.1.sequence                             46875   \n",
       "550.L1S20.s.1.sequence                              30094   \n",
       "\n",
       "                         observations_deblur_90bp  observations_deblur_100bp  \n",
       "#SampleID                                                                     \n",
       "550.L1S116.s.1.sequence                     22567                      22160  \n",
       "550.L1S119.s.1.sequence                     27871                      27191  \n",
       "550.L1S164.s.1.sequence                     24134                      23686  \n",
       "550.L1S194.s.1.sequence                     30041                      29264  \n",
       "550.L1S20.s.1.sequence                      21132                      20643  "
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:,17:23].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64789.0535"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seqs.observations_deblur_90bp.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
